src/h3/mod.rs

```diff
@@ -1080,22 +1108,45 @@ impl Connection {
             return Ok((finished, Event::Finished));
         }
 
-        // Process DATAGRAMs
-        let mut d = [0; 8];
+        // Process DATAGRAMs.
+        let process_dgram = match self.dgram_poll_threshold {
+            None => true,
 
-        match conn.dgram_recv_peek(&mut d, 8) {
-            Ok(_) => {
-                let mut b = octets::Octets::with_slice(&d);
-                let flow_id = b.get_varint()?;
-                return Ok((flow_id, Event::Datagram));
-            },
+            Some(thresh) => self.dgram_poll_count < thresh,
+        };
 
-            Err(crate::Error::Done) => (),
+        if process_dgram {
+            let mut d = [0; 8];
 
-            Err(e) => return Err(Error::TransportError(e)),
-        };
+            match conn.dgram_recv_peek(&mut d, 8) {
+                Ok(_) => {
+                    if self.dgram_poll_threshold.is_some() {
+                        self.dgram_poll_count += 1;
+                    }
+
+                    let mut b = octets::Octets::with_slice(&d);
+                    let flow_id = b.get_varint()?;
+                    return Ok((flow_id, Event::Datagram));
+                },
+
+                Err(crate::Error::Done) => (),
+
+                Err(e) => return Err(Error::TransportError(e)),
+            };
+        }
 
         // Process HTTP/3 data from readable streams.
+        if let Some(thresh) = self.stream_poll_threshold {
+            if self.stream_poll_count > thresh {
+                self.dgram_poll_count = 0;
+                self.stream_poll_count = 0;
+
+                return Err(Error::Done);
+            }
+
+            self.stream_poll_count += 1;
```

Comment:
I think this should be inside the loop, and only increased if we actually return a stream event. Otherwise it seems like it would be increased even when there are no stream events and `Done` is returned?

Reply:
Yeah this seems a bit wonky but during testing the problem I hit is that once you're into processing streams, you have to have a stream event in order to get back to datagrams - but that might never happen e,g, a basic MASQUE session.

The current code is intended to say "I'll try to process only stream events for N calls of poll, after that I'll go back to trying datagrams". It's not a count of the number of events, its the poll count.

Reply:
Normally when `poll()` returns `Done` the application will stop calling it until, e.g. new packets arrive or something. So this behavior might end up delaying the delivery of DATAGRAM events for several RTTs I think.

Instead of doing that, when checking if DATAGRAM can be sent, we can also call `conn.streams.has_readable()` to see if there even are any streams that are readable. so e.g. `if !conn.streams.has_readable() || below dgram threshold {` or something like that.

Reply:
I hadn't appreciated the importance of `Done` before. This causes a mild quandry because there is no outright correct order. Ideally, while there is any DATAGRAM or stream data buffered and ready to be processed, we want poll to return an event even if that "bends" the yield thresholds. We really don't want to have call `poll()` multiple times and have it return `Done` in order to get over the threshold to start processing the data we care about. 

I've tried to address this in https://github.com/cloudflare/quiche/pull/704/commits/91eab5f0dc1ac461e20c84a7e0baf48d329d073e. With that change, the intent is that `poll()` will jump into processing whatever type the connection-global threshold prefers. If there is nothing to do for a given type, switch to the other one. It only returns `Done` when there is nothing to do for either type.